smart_open/gcs.py

```diff
@@ -141,28 +125,34 @@ def seek(self, position):
         """
         self._position = position
 
-        if position == self._size == 0 or position == self._size:
+    def read(self, size=-1):
+        if self._position >= self._size:
+            return b''
+        binary = self._download_blob_chunk(size)
+        self._position += len(binary)
+        return binary
+
+    def _download_blob_chunk(self, size):
+        position = self._position
+        if position == self._size:
             #
             # When reading, we can't seek to the first byte of an empty file.
             # Similarly, we can't seek past the last byte.  Do nothing here.
             #
-            self._body = io.BytesIO()
-        else:
-            start, end = position, position + self._size
-            self._body = self._blob.download_as_string(start=start, end=end)
-
-    def read(self, size=-1):
-        if self._position >= self._size:
-            return b''
-        if size == -1:
-            binary = self._body
+            binary = b''
+        elif size == -1:
+            binary = self._blob.download_as_string(start=position)
         else:
-            binary = self._body[:size]
-        self._position += len(binary)
+            start, end = position, position + size
+            binary = self._blob.download_as_string(start=start, end=end)
```

Comment:
What happens when trying to read past the end of the file? `end > self._size`

Reply:
`google.cloud.storage.Blob.download_as_string` appears to handle this for us. I added a test to verify.

```python
@maybe_mock_gcs
    def test_read_past_end(self):
        content = b'englishman\nin\nnew\nyork\n'
        put_to_bucket(contents=content)

        with smart_open.gcs.SeekableBufferedInputBase(BUCKET_NAME, BLOB_NAME) as fin:
            data = fin.read(100)

        self.assertEqual(data, content)
```